floodmodeller_api/toolbox/model_build/structure_log/structure_log.py

```diff
@@ -272,16 +514,72 @@ def _write(  # noqa: PLR0913
             ],
         )
 
+    def _write_csv_output(self, file):
+        """
+        Take the current state of the instance (unit_store etc) and write it to the specified output file.
+        """
+        
+        writer = csv.writer(file)
+
+        self._add_fields(writer)
+
+        for _, unit_dict in self.unit_store.items():
+            name = unit_dict["name"]
+            unit_type = unit_dict["type"]
+            subtype = unit_dict["subtype"]
+            comment = unit_dict["comment"]
+
+            friction = self._format_friction(unit_dict)
+
+            culvert_loss = ""
+
+            match (unit_type):
+                case "BRIDGE":
+                    dimensions = self._format_bridge_dimensions(unit_dict)
+                case "ORIFICE":
+                    dimensions = self._format_orifice_dimensions(unit_dict)
+                case "WEIR" | "RNWEIR":
+                    dimensions = self._format_weir_dimensions(unit_dict)
+                case "SLUICE":
+                    dimensions = self._format_sluice_dimensions(unit_dict)
+                case "SPILL":
+                    dimensions = self._format_spill_dimensions(unit_dict)
+                case "CONDUIT":
+                    dimensions, culvert_loss = self._format_conduit_dimensions(unit_dict)
+                case _:
+                    dimensions = ""
+
+            try:
+                weir_coefficient = unit_dict["dimensions"]["weir_coefficient"]
+            except KeyError:
+                weir_coefficient = ""
+
+            self._write(
+                writer,
+                name,
+                unit_type,
+                subtype,
+                comment,
+                friction,
+                dimensions,
+                weir_coefficient,
+                culvert_loss,
+            )
+
     def create(self):
-        # Read in the .dat file
+        """
+        When using the toolbox wrapper or commandline entry point, this is the entrypoint to the structure logger code
+        """
         self._dat = DAT(self.dat_file_path)
-
-        # Create a new .csv file
+        self._add_conduits()
+        self._add_structures()
         with open(self.csv_output_path, "w", newline="") as file:
-            self._writer = csv.writer(file)
-
-            self._add_fields()
+            self._write_csv_output(file)
 
-            self._add_conduits()
 
-            self._add_structures()
+if __name__ == "__main__":
```

Comment:
I would remove this before merging, but if it is an informal test, then you can turn it into a test if you want

Reply:
deleted and moved to off-repo file